### PR TITLE
Allow `limit_type_depth` to introduce more than one new `TypeVar`

### LINKED ISSUE #20615: zip() has limits: bug or needs documenting?
Travis found this bug for me on 0.6: It seems that `zip()` can handle up to a certain number of inputs, then complains.

```
julia> a = 1:5
1:5

julia> b = ["e","d","b","c","a"]
5-element Array{String,1}:
 "e"
 "d"
 "b"
 "c"
 "a"

julia> for e in zip(a, b)
           println(e)
       end
(1, "e")
(2, "d")
(3, "b")
(4, "c")
(5, "a")

julia> for e in zip(a, b, a, b, a, b, a, b, a, b, a)
           println(e)
       end
(1, "e", 1, "e", 1, "e", 1, "e", 1, "e", 1)
(2, "d", 2, "d", 2, "d", 2, "d", 2, "d", 2)
(3, "b", 3, "b", 3, "b", 3, "b", 3, "b", 3)
(4, "c", 4, "c", 4, "c", 4, "c", 4, "c", 4)
(5, "a", 5, "a", 5, "a", 5, "a", 5, "a", 5)

julia> for e in zip(a, b, a, b, a, b, a, b, a, b, a, b)
           println(e)
       end
ERROR: type Array has no field a
Stacktrace:
 [1] start at ./iterators.jl:161 [inlined] (repeats 10 times)
 [2] anonymous at ./<missing>:?
```

# Version

```
julia> versioninfo()
Julia Version 0.6.0-dev.2635
Commit dc2459d (2017-02-13 07:46 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin13.4.0)
  CPU: Intel(R) Core(TM) i5-2500S CPU @ 2.70GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, sandybridge)

```

### PR DESCRIPTION
**Pull request by martinholters**
Fixes #20615.

~This is WIP because it increases the time to run the `fft` test from 20s to 630s on my system. (All other tests look unsuspicious at first glance.)~

### CODE REVIEW

**Comment by martinholters** [comment]
Looks like the `foldr` was the culprit. Spelling the loop out is vastly faster (at least for the fft test). No idea what's special about that test and why `foldr` is so slow there.

**Comment by martinholters** [comment]
I wonder whether types deep enough for this PR to make a difference appear in the benchmarks, but better be safe than sorry, so @nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by martinholters** [comment]
Should we do something like
```
--- a/base/inference.jl
+++ b/base/inference.jl
@@ -660,7 +660,13 @@ function type_depth(t::ANY)
     return 0
 end
 
-function limit_type_depth(t::ANY, d::Int, cov::Bool=true, vars::Vector{TypeVar}=TypeVar[])
+function limit_type_depth(t::ANY, d::Int)
+    r = limit_type_depth(t, d, true, TypeVar[])
+    @assert t <: r
+    return r
+end
+
+function limit_type_depth(t::ANY, d::Int, cov::Bool, vars::Vector{TypeVar}=TypeVar[])
     if isa(t,Union)
         if d > MAX_TYPE_DEPTH
             return Any
```
to turn potentially very subtle bugs into more obvious ones? (Of course, `limit_type_depth` will be absolutely free of any bugs with this PR, but just to be sure... :smile:)

**Comment by JeffBezanson** [comment]
Nice, :+1:.

Only one small thing; I would not use gensym here since it might generate a large number of symbols that aren't strictly necessary. Yes, using separate names makes the types easier to read, but we can use a symbol pool instead.

**Comment by martinholters** [comment]
I'd like to do something like
```Julia
var = TypeVar(Symbol("_", length(vars) + 1), t.name.wrapper)
```
but string concatenation is not available during inference. Is there any easy way to accomplish this? With a "symbol pool", you probably meant having something like `[:_1, :_2, ..., :_9]` around and then cycling through it to achieve basically the same?

My current thinking is: Either go all the way and ensure that the `TypeVar`s have different `Symbol`s so that they parse as they print - but without `gensym()` that might be more complicated than is justified, or just use `:_` throughout, which I've done now.

**Comment by TotalVerb** [comment]
Scott reported a problem, possibly related to this (as it seems to trigger the assertion here), on Discourse: https://discourse.julialang.org/t/strange-failure-in-v0-6/2911/8

File: base/inference.jl [hunk 1]
```diff
@@ -660,21 +660,27 @@ function type_depth(t::ANY)
...
+function limit_type_depth(t::ANY, d::Int)
+    r = limit_type_depth(t, d, true, TypeVar[])
+    @assert !isa(t, Type) || t <: r
+    return r
+end
+
+function limit_type_depth(t::ANY, d::Int, cov::Bool, vars::Vector{TypeVar}=TypeVar[])
     if isa(t,Union)
         if d > MAX_TYPE_DEPTH
             return Any
```
Comment (TotalVerb) [inline base/inference.jl]: Should this have been `cov ? Any : (v = TypeVar(:_); push!(vars, v); v)`? cc @martinholters

Comment (martinholters) [inline base/inference.jl]: Sounds reasonable. I'll look into this...

Comment (martinholters) [inline base/inference.jl]: PR at #21192, but also needs #21191 fixed.

**Pull request merged**